internal/controller/iprange_controller.go

```diff
@@ -0,0 +1,306 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const IpRangeFinalizerName = "iprange.netbox.dev/finalizer"
+const LastIpRangeMetadataAnnotationName = "iprange.netbox.dev/last-ip-range-metadata"
+
+// IpRangeReconciler reconciles a IpRange object
+type IpRangeReconciler struct {
+	client.Client
+	Scheme            *runtime.Scheme
+	NetboxClient      *api.NetboxClient
+	Recorder          record.EventRecorder
+	OperatorNamespace string
+	RestConfig        *rest.Config
+}
+
+//+kubebuilder:rbac:groups=netbox.dev,resources=ipranges,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=netbox.dev,resources=ipranges/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=netbox.dev,resources=ipranges/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *IpRangeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("reconcile loop started")
+
+	o := &netboxv1.IpRange{}
+	err := r.Client.Get(ctx, req.NamespacedName, o)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// if being deleted
+	if !o.ObjectMeta.DeletionTimestamp.IsZero() {
+		if o.Spec.PreserveInNetbox {
+			// if there's a finalizer remove it and return
+			// this can be the case if a CR used to have spec.preserveInNetbox set to false
+			return ctrl.Result{}, r.removeFinalizer(ctx, o)
+		}
+
+		err := r.NetboxClient.DeleteIpRange(o.Status.IpRangeId)
+		if err != nil {
+			err = r.logErrorSetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeReadyFalseDeletionFailed,
+				corev1.EventTypeWarning, "", err)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		err = r.removeFinalizer(ctx, o)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		err = r.Update(ctx, o)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// if PreserveIpInNetbox flag is false then register finalizer if not yet registered
+	if !o.Spec.PreserveInNetbox && !controllerutil.ContainsFinalizer(o, IpRangeFinalizerName) {
+		logger.V(4).Info("adding the finalizer")
+		controllerutil.AddFinalizer(o, IpRangeFinalizerName)
+		if err := r.Update(ctx, o); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// 1. try to lock lease of parent prefix if IpRange status condition is not true
+	// and IpRange is owned by an IpRangeClaim
+	or := o.ObjectMeta.OwnerReferences
+	var ll *leaselocker.LeaseLocker
+	if len(or) > 0 /* len(nil array) = 0 */ && !apismeta.IsStatusConditionTrue(o.Status.Conditions, "Ready") {
```

Comment:
Let's remove this comment `/* len(nil array) = 0 */`. It's something that I somehow left behind when fixing the linter error long ago... Sorry

```suggestion
	if len(or) > 0 && !apismeta.IsStatusConditionTrue(o.Status.Conditions, "Ready") {
```